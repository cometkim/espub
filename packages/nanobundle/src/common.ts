export type PathResolver = (...paths: string[]) => string;
export type RelativePathResolver = (path: string, startsWithDot?: boolean) => string;
