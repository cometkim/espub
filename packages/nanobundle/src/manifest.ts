import * as path from 'node:path';
import * as fs from 'node:fs/promises';

export type Manifest = {
  name?: string,

  version?: string,

  type?: 'commonjs' | 'module',

  /**
   * Source file for the `main`, `module`, and `exports` entry
   */
  source?: string,

  // Non-standard entry style for legacy bundlers
  module?: string,

  // Main entry
  main?: string,

  // Binary entries
  bin?: string | {
    [name: string]: string,
  },

  // TypeScript declaration for "main" entry
  types?: string,

  // Export maps
  exports?: ConditionalExports,

  // Subpath imports
  imports?: ConditionalImports,

  dependencies?: {
    [name: string]: string,
  },

  peerDependencies?: {
    [name: string]: string,
  },

  browserslist?: string | string[],

  engines?: {
    node?: string,
    deno?: string,
  },
};

// See https://nodejs.org/api/packages.html#packages_nested_conditions
// What a mess :/
export type ConditionalExports = (
  | string
  | {
    [module: string]: ConditionalExports,
  }
  | {
    'import'?: ConditionalExports,
    'require'?: ConditionalExports,
    'node'?: ConditionalExports,
    'node-addons'?: ConditionalExports,
    'default'?: ConditionalExports,

    // community conditions definitions
    // See https://nodejs.org/api/packages.html#packages_community_conditions_definitions
    // See also https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#package-json-exports-imports-and-self-referencing
    'types'?: ConditionalExports,
    'deno'?: ConditionalExports,
    'browser'?: ConditionalExports,
    'development'?: ConditionalExports,
    'production'?: ConditionalExports,
  }
);

export type ConditionalImports = (
  | string
  | {
    [module: string]: ConditionalImports,
  }
  | {
    'import'?: ConditionalImports,
    'require'?: ConditionalImports,
    'node'?: ConditionalImports,
    'default'?: ConditionalImports,
    'browser'?: ConditionalImports,
    'development'?: ConditionalImports,
    'production'?: ConditionalImports,
  }
);

type ManifestWithOverride = Manifest & {
  publishConfig?: Manifest,
};

interface LoadManifest {
  (props: {
    basePath: string,
  }): Promise<Manifest>;
}
export const loadManifest: LoadManifest = async ({
  basePath,
}) => {
  const configPath = path.resolve(basePath, 'package.json');

  const { publishConfig, ...config } = await fs.readFile(configPath, 'utf-8')
    .then(JSON.parse) as ManifestWithOverride;

  return {
    ...config,
    ...publishConfig,
  };
}
