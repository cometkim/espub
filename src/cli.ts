#!/usr/bin/env node

import meow from 'meow';

export const cli = meow(`
Usage
  $ nanobundle <command> [options]

Available Commands
  build    Build once and exit

Options
  --version          Display current version

  --cwd              Use an alternative working directory

  --root-dir         Specify the path to resolve source entry (default: ./src)
                     This can be overriden by tsconfig.json

  --out-dir          Specify the path to resolve source entry (default: .)
                     This can be overriden by tsconfig.json

  --tsconfig         Specify the path to a custom tsconfig.json

  --import-maps      Specify import map file path  (default: package.json)

  --standalone       Embed external dependencies into the bundle (default: false)

  --external         Specify external dependencies to exclude from the bundle

  --minify           Minify production output (default: true)

  --sourcemap        Generate source map (default: true)

  --dts              Generate TypeScript .d.ts files (default: true)

  --help             Display this message
`, {
  importMeta: import.meta,
  flags: {
    cwd: {
      type: 'string',
      default: process.cwd(),
    },
    rootDir: {
      type: 'string',
      default: 'src',
    },
    outDir: {
      type: 'string',
      default: '.',
    },
    tsconfig: {
      type: 'string',
      default: 'tsconfig.json',
    },
    importMaps: {
      type: 'string',
      default: 'package.json',
    },
    external: {
      type: 'string',
      isMultiple: true,
      default: [],
    },
    standalone: {
      type: 'boolean',
      default: false,
    },
    minify: {
      type: 'boolean',
      default: true,
    },
    sourcemap: {
      type: 'boolean',
      default: true,
    },
    dts: {
      type: 'boolean',
      default: true,
    },
  },
});

export type Flags = typeof cli.flags;
