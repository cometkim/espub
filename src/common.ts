export type PathResolver = (cwd: string, path: string) => string;