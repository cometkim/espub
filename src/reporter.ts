import { formatWithOptions } from 'node:util';
import kleur from 'kleur';

import * as formatUtils from './formatUtils';
import { NanobundleError } from './errors';

export interface Reporter {
  debug(msg: string, ...args: any[]): void;
  info(msg: string, ...args: any[]): void;
  warn(msg: string, ...args: any[]): void;
  error(msg: string, ...args: any[]): void;
  captureException(exn: unknown): void;
  createChildReporter(): Reporter;
}

export class ConsoleReporter implements Reporter {
  #level: number;
  #console: Console;

  color = formatUtils.colorEnabled;
  level: 'default' | 'debug' = 'debug';

  constructor(console: Console, level = 0) {
    this.#level = level;
    this.#console = console;
  }

  #indent(msg: string): string {
    return formatUtils.indent(msg, this.#level);
  }

  debug(msg: string, ...args: any[]): void {
    if (this.level !== 'debug') {
      return;
    }

    const formatted = formatWithOptions(
      { colors: this.color },
      msg,
      ...args,
    );
    const indented = this.#indent(formatted);
    this.#console.debug(
      kleur.gray(`[debug] ${indented}`),
    );
  }

  info(msg: string, ...args: any[]): void {
    const formatted = formatWithOptions(
      { colors: this.color },
      msg,
      ...args,
    );
    const indented = this.#indent(formatted);
    this.#console.info(
      kleur.white(`[info] ${indented}`),
    );
  }

  warn(msg: string, ...args: any[]): void {
    const formatted = formatWithOptions(
      { colors: this.color },
      msg,
      ...args,
    );
    const indented = this.#indent(formatted);
    this.#console.warn(
      kleur.yellow(`[warn] ${indented}`),
    );
  }

  error(msg: string, ...args: any[]): void {
    const formatted = formatWithOptions(
      { colors: this.color },
      msg,
      ...args,
    );
    const indented = this.#indent(formatted);
    this.#console.error(
      kleur.red(`[error] ${indented}`),
    );
  }

  captureException(exn: unknown): void {
    let formatted;
    if (exn instanceof NanobundleError && exn.message) {
      formatted = exn.message;
    } else if (exn instanceof Error) {
      formatted = formatWithOptions(
        { colors: this.color },
        exn.stack,
      );
    } else {
      formatted = formatWithOptions(
        { colors: this.color },
        '%s',
        exn,
      );
    }
    const indented = this.#indent(formatted);
    this.#console.error(
      kleur.bold().red(`${indented}`),
    );
  }

  createChildReporter(): ConsoleReporter {
    const child = new ConsoleReporter(this.#console, this.#level + 1);
    child.color = this.color;
    return child;
  }
};
