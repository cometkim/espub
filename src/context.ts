import dedent from 'string-dedent';
import { type TSConfig } from 'pkg-types';
import * as semver from 'semver';

import { type Flags } from './cli';
import { type Manifest } from './manifest';
import { type Entry } from './entry';
import { type Reporter } from './reporter';
import { type PathResolver } from './common';
import * as formatUtils from './formatUtils';

export class NanobundleConfigError extends Error {
  name = 'NanobundleConfigError';
}

export type Context = {
  cwd: string,
  module: Entry['module'],
  platform: Entry['platform'],
  sourcemap: boolean,
  declaration: boolean,
  standalone: boolean,
  rootDir: string,
  outDir: string,
  tsconfigPath: string,
  importMapsPath: string,
  externalDependencies: string[],
  forceExternalDependencies: string[],
  manifest: Manifest,
  targets: string[],
  reporter: Reporter,
  resolve: PathResolver,
};

export type Config = {
  flags: Flags,
  manifest: Manifest,
  targets: string[],
  reporter: Reporter,
  resolve: PathResolver,
  tsconfig?: TSConfig,
  tsconfigPath?: string,
};

export function parseConfig({
  flags,
  manifest,
  targets: inputTargets,
  reporter,
  resolve,
  tsconfig,
  tsconfigPath: resolvedTsConfigPath,
}: Config): Context {
  const cwd = flags.cwd;
  const sourcemap = !flags.noSourcemap;
  const standalone = flags.standalone;
  const tsconfigPath = resolvedTsConfigPath || flags.tsconfig;
  const importMapsPath = flags.importMaps;
  const forceExternalDependencies = flags.external;
  const externalDependencies = [
    ...(manifest.dependencies ? Object.keys(manifest.dependencies) : []),
    ...(manifest.peerDependencies ? Object.keys(manifest.peerDependencies) : []),
    ...forceExternalDependencies,
  ];

  const rootDir: string = (
    flags.rootDir ||
    tsconfig?.compilerOptions?.rootDir ||
    'src'
  );
  const outDir: string = (
    flags.outDir ||
    tsconfig?.compilerOptions?.outDir ||
    'lib'
  );

  const module = (
    manifest.type === 'module'
      ? 'esmodule'
      : 'commonjs'
  );

  let platform: Entry['platform'] = 'neutral';
  if (['node', 'deno', 'web'].includes(flags.platform || '')) {
    platform = flags.platform as Entry['platform'];
  } else if (manifest.engines?.node) {
    platform = 'node';
  }

  let targets = [...inputTargets];
  if (manifest.engines?.node) {
    const version = semver.minVersion(manifest.engines.node);
    if (version) {
      targets = [`node${version.major}`];
    }
  }
  if (platform === 'node' && !targets.some(target => target.startsWith('node'))) {
    targets = ['node14'];
  }

  let declaration = false;
  if (!flags.noDts && tsconfig) {
    declaration = tsconfig.compilerOptions?.declaration !== false;
  }

  if (!declaration && rootDir === outDir) {
    throw new NanobundleConfigError(dedent`
      ${formatUtils.key('rootDir')} (${formatUtils.path(rootDir)}) and ${formatUtils.key('outDir')} (${formatUtils.path(outDir)}) are conflict!

      Please specify different directory for one of them.
    `,)
  }

  return {
    cwd,
    module,
    platform,
    sourcemap,
    declaration,
    standalone,
    rootDir,
    outDir,
    tsconfigPath,
    importMapsPath,
    externalDependencies,
    forceExternalDependencies,
    manifest,
    targets,
    reporter,
    resolve,
  };
};
