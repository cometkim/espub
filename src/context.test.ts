import { describe, test, expect, vi } from 'vitest';
import { type TSConfig } from 'pkg-types';

import { type PathResolver } from './common';
import { type Context } from './context';
import { type Flags } from './cli';
import { type Manifest } from './manifest';
import { type Reporter } from './reporter';
import { parseConfig } from './context';

class ViReporter implements Reporter {
  debug = vi.fn();
  info = vi.fn();
  warn = vi.fn();
  error = vi.fn();
  captureException = vi.fn();
  createChildReporter() {
    return new ViReporter();
  }
}

describe('parseConfig', () => {
  const reporter = new ViReporter();
  const resolve: PathResolver = vi.fn();
  const defaultFlags: Flags = {
    cwd: '/project',
    verbose: false,
    platform: undefined,
    rootDir: undefined,
    outDir: undefined,
    tsconfig: 'tsconfig.json',
    importMaps: 'package.json',
    jsx: undefined,
    jsxFactory: undefined,
    jsxFragment: undefined,
    jsxImportSource: undefined,
    external: [],
    standalone: false,
    dts: true,
    sourcemap: true,
  };
  const defaultManifest: Manifest = {
    name: 'package'
  };
  const defaultTsConfig: TSConfig = {
    compilerOptions: {
      target: 'ESNext', // ESNext,
      declaration: true,
    },
  };
  const defaultTargets: string[] = [
    'chrome',
    'firefox',
    'safari',
  ];

  test('validate manifest', () => {
    const result = parseConfig({
      flags: defaultFlags,
      manifest: defaultManifest,
      targets: defaultTargets,
      reporter,
      resolve,
    });

    expect(result).toEqual<Context>({
      cwd: '/project',
      verbose: false,
      module: 'commonjs',
      platform: 'neutral',
      sourcemap: true,
      declaration: false,
      standalone: false,
      jsx: undefined,
      jsxDev: false,
      jsxFactory: 'React.createElement',
      jsxFragment: 'Fragment',
      jsxImportSource: 'react',
      rootDir: 'src',
      outDir: 'lib',
      tsconfigPath: undefined,
      importMapsPath: 'package.json',
      externalDependencies: [],
      forceExternalDependencies: [],
      manifest: defaultManifest,
      targets: defaultTargets,
      reporter,
      resolve,
    });
  });

  describe('flags', () => {
    test('--rootDir', () => {
      const result = parseConfig({
        flags: {
          ...defaultFlags,
          rootDir: '.',
        },
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: false,
        standalone: false,
        rootDir: '.',
        outDir: 'lib',
        jsx: undefined,
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        tsconfigPath: undefined,
        importMapsPath: 'package.json',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('--outDir', () => {
      const result = parseConfig({
        flags: {
          ...defaultFlags,
          outDir: '.',
        },
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: false,
        standalone: false,
        rootDir: 'src',
        outDir: '.',
        tsconfigPath: undefined,
        jsx: undefined,
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        importMapsPath: 'package.json',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('rootDir=outDir is not allowed without TypeScript', () => {
      expect(() => parseConfig({
        flags: {
          ...defaultFlags,
          rootDir: '.',
          outDir: '.',
        },
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      })).toThrowErrorMatchingSnapshot();
    });
  });

  describe('node platform', () => {
    test('platform should be node when engines.node exist', () => {
      const manifest = {
        ...defaultManifest,
        engines: {
          node: '>=16.0.0',
        },
      };

      const result = parseConfig({
        flags: defaultFlags,
        targets: defaultTargets,
        manifest,
        reporter,
        resolve,
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'node',
        sourcemap: true,
        declaration: false,
        standalone: false,
        rootDir: 'src',
        outDir: 'lib',
        tsconfigPath: undefined,
        importMapsPath: 'package.json',
        jsx: undefined,
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest,
        targets: [
          'node16',
        ],
        reporter,
        resolve,
      });
    });

    test('default node version is 14', () => {
      const result = parseConfig({
        flags: {
          ...defaultFlags,
          platform: 'node',
        },
        targets: defaultTargets,
        manifest: defaultManifest,
        reporter,
        resolve,
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'node',
        sourcemap: true,
        declaration: false,
        standalone: false,
        rootDir: 'src',
        outDir: 'lib',
        tsconfigPath: undefined,
        importMapsPath: 'package.json',
        jsx: undefined,
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: [
          'node14',
        ],
        reporter,
        resolve,
      });
    });
  });

  describe('externalDependencies', () => {
    test.todo('externalDependencies has manifest dependencies')
    test.todo('forceExternalDependencies always include --external flag list')
    test.todo('externalDependencies always include --external flag list')
  });

  describe('tsCompilerOptions', () => {
    test('declaration=true', () => {
      const result = parseConfig({
        flags: defaultFlags,
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
        tsconfigPath: 'tsconfig.json',
        tsconfig: defaultTsConfig,
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: true,
        standalone: false,
        rootDir: 'src',
        outDir: 'lib',
        tsconfigPath: 'tsconfig.json',
        importMapsPath: 'package.json',
        jsx: undefined,
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('declaration=false', () => {
      const result = parseConfig({
        flags: defaultFlags,
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
        tsconfigPath: 'tsconfig.json',
        tsconfig: {
          compilerOptions: {
            ...defaultTsConfig.compilerOptions,
            declaration: false,
            standalone: false,
          },
        },
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: false,
        standalone: false,
        rootDir: 'src',
        outDir: 'lib',
        tsconfigPath: 'tsconfig.json',
        importMapsPath: 'package.json',
        jsx: undefined,
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('declaration=true, --no-dts', () => {
      const result = parseConfig({
        flags: {
          ...defaultFlags,
          dts: false,
        },
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
        tsconfigPath: 'tsconfig.json',
        tsconfig: defaultTsConfig,
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: false,
        standalone: false,
        rootDir: 'src',
        outDir: 'lib',
        tsconfigPath: 'tsconfig.json',
        importMapsPath: 'package.json',
        jsx: undefined,
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('respect rootDir and outDir in compilerOptions', () => {
      const result = parseConfig({
        flags: defaultFlags,
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
        tsconfigPath: 'tsconfig.json',
        tsconfig: {
          ...defaultTsConfig,
          compilerOptions: {
            ...defaultTsConfig.compilerOptions,
            rootDir: 'tsconfig-src',
            outDir: 'tsconfig-lib',
          },
        },
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: true,
        standalone: false,
        rootDir: 'tsconfig-src',
        outDir: 'tsconfig-lib',
        tsconfigPath: 'tsconfig.json',
        importMapsPath: 'package.json',
        jsx: undefined,
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('respect jsx in compilerOptions', () => {
      const result = parseConfig({
        flags: defaultFlags,
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
        tsconfigPath: 'tsconfig.json',
        tsconfig: {
          ...defaultTsConfig,
          compilerOptions: {
            ...defaultTsConfig.compilerOptions,
            jsx: 'react-jsxdev',
          },
        },
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: true,
        standalone: false,
        rootDir: 'src',
        outDir: 'lib',
        tsconfigPath: 'tsconfig.json',
        importMapsPath: 'package.json',
        jsx: 'automatic',
        jsxDev: true,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('respect jsxSource in compilerOptions', () => {
      const result = parseConfig({
        flags: defaultFlags,
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
        tsconfigPath: 'tsconfig.json',
        tsconfig: {
          ...defaultTsConfig,
          compilerOptions: {
            ...defaultTsConfig.compilerOptions,
            jsx: 'react',
            jsxFactory: 'h',
            jsxFragmentFactory: 'Fragment',
            jsxImportSource: 'preact',
          },
        },
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: true,
        standalone: false,
        rootDir: 'src',
        outDir: 'lib',
        tsconfigPath: 'tsconfig.json',
        importMapsPath: 'package.json',
        jsx: 'transform',
        jsxDev: false,
        jsxFactory: 'h',
        jsxFragment: 'Fragment',
        jsxImportSource: 'preact',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('flags precedense over tsconfig', () => {
      const result = parseConfig({
        flags: {
          ...defaultFlags,
          jsx: 'preserve',
          rootDir: 'flags-src',
          outDir: 'flags-lib',
        },
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
        tsconfigPath: 'tsconfig.json',
        tsconfig: {
          ...defaultTsConfig,
          compilerOptions: {
            ...defaultTsConfig.compilerOptions,
            jsx: 'react-jsxdev',
            rootDir: 'tsconfig-src',
            outDir: 'tsconfig-lib',
          },
        },
      });

      expect(result).toEqual<Context>({
        cwd: '/project',
        verbose: false,
        module: 'commonjs',
        platform: 'neutral',
        sourcemap: true,
        declaration: true,
        standalone: false,
        rootDir: 'flags-src',
        outDir: 'flags-lib',
        tsconfigPath: 'tsconfig.json',
        importMapsPath: 'package.json',
        jsx: 'preserve',
        jsxDev: false,
        jsxFactory: 'React.createElement',
        jsxFragment: 'Fragment',
        jsxImportSource: 'react',
        externalDependencies: [],
        forceExternalDependencies: [],
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
      });
    });

    test('rootDir=outDir is allowed with TypeScript', () => {
      expect(() => parseConfig({
        flags: {
          ...defaultFlags,
          rootDir: '.',
          outDir: '.',
        },
        manifest: defaultManifest,
        targets: defaultTargets,
        reporter,
        resolve,
        tsconfig: defaultTsConfig,
      })).not.toThrowError();
    });
  });
});
