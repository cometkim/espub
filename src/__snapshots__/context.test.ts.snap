// Vitest Snapshot v1

exports[`parseConfig > flags > rootDir=outDir is not allowed without TypeScript 1`] = `
"\\"rootDir\\" (.) and \\"outDir\\" (.) are conflict!

Please specify different directory for one of them."
`;
